### DISCUSSION TITLE
got "AttributeError" when run tox

### DISCUSSION THREAD

**Discussion by mmdbalkhi** [tox-dev/tox#2512] [Q&A] [answered]
I recently moved my projects to use tox. when I run the `tox` I got ` AttributeError: module 'flask_tinydb' has no attribute '__version__'` but the `pytest` command didn't return this. how I can fix it?

tox file: https://github.com/mmdbalkhi/Flask-tinydb/blob/main/tox.ini
`__version__` att: https://github.com/mmdbalkhi/Flask-tinydb/blob/fb4e09d3d225f3a49ee7b2b663e79e1b7f7a8d01/flask_tinydb/__init__.py#L34

Thank you in advance for your answer.

**Comment by jugmac00** [comment, marked as answer]
Before `tox` runs tests, it builds the package, installs it and runs the test against the installed version, not the files in your source tree (the git checkout). This is a good thing, as usually that is what your users use.

When running `check-manifest` on your package, it is clear that your `MANIFEST.in` is not correct, and so a couple of files are not included in the `sdist`:

```
Flask-tinydb on  main via 🐍 v2.7.18 
❯ check-manifest
lists of files in version control and sdist do not match!
missing from sdist:
  .editorconfig
  .pre-commit-config.yaml
  flask_tinydb/__init__.py
  flask_tinydb/storages.py
  flask_tinydb/tinydb.py
  requirements/dev.in
  requirements/dev.txt
  requirements/docs.in
  requirements/docs.txt
  requirements/tests.in
  requirements/tests.txt
  requirements/typing.in
  requirements/typing.txt
  tox.ini
suggested MANIFEST.in rules:
  include *.yaml
  include tox.ini
  recursive-include flask_tinydb *.py
  recursive-include requirements *.in
  recursive-include requirements *.txt
```

This is the one issue. Unfortunately, it is not fixed by just adding the above instructions.

In your `requirements/tests.txt` (and so also in your requirements/tests.in) `pytest` is missing.

In general, it is recommended that you use a `src` directory as the top level directory for your package, which then contains no `__init__.py` so Python cannot import from it. Otherwise there is the issue.. when you write `from flask_tinydb import TinyDB`... does it import from the file `flask_tinydb.py` or from the installed package?

I recommend to read the following articles:
https://hynek.me/articles/testing-packaging/
https://web.archive.org/web/20170615032800/https://enotuniq.org/
https://blog.ganssle.io/articles/2019/08/test-as-installed.html
